Add integration for Hegel Music Systems amplifiers

boazca opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new integration for **Hegel Music Systems amplifiers**, enabling Home Assistant users to control and monitor their Hegel amplifiers directly from Home Assistant.

**Key Features:**

*   **Real-time control**: Power, volume, mute, and source selection with instant feedback
*   **Push notifications**: Immediate updates when changes are made via front panel or remote
*   **Automatic discovery**: SSDP/UPnP discovery for seamless setup
*   **Connection resilience**: Automatic reconnection with exponential backoff
*   **Model-specific input mapping**: Proper input names for different Hegel models (H95, H120, H190, H190V, H390, H590, Röst)
*   **Advanced services**: Raw command service for power users
*   **Comprehensive error handling**: Robust connection management and recovery

**Supported Models:**

*   Hegel Röst, H95, H120, H190, H190V, H390, H590
*   Other Hegel models with IP control support may also work

The integration uses Hegel's official TCP/IP control protocol (port 50001) and achieves **Silver quality scale** with full config flow support, proper entity unique IDs, connection validation during setup, runtime data storage, and comprehensive test coverage.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/41210
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

boazca committed:
Add Hegel integration

boazca committed:
Add strings.json

boazca committed:
Merge branch 'home-assistant:dev' into dev

boazca committed:
Merge branch 'home-assistant:dev' into dev

boazca committed:
Improve handling of connection reset and multiple connections

boazca committed:
fix client to establish connection once

boazca committed:
fix typo in README.md

boazca committed:
update integrations to align with bronze standard

boazca committed:
update integrations to align with silver standard

boazca committed:
Merge branch 'home-assistant:dev' into dev

boazca committed:
format using `ruff format homeassistant tests`

boazca committed:
Merge branch 'dev' into dev

pantherale0 commented on homeassistant/components/hegel/hegel_client.py:
```diff
@@ -0,0 +1,284 @@
+CR = "\r"
+
+
+class HegelClient:
```
I think this, along with a number of the consts should be moved to an external Python module.

joostlek replied:
Yes, device specific code should be hosted in a python library

pantherale0 commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,395 @@
+    }
+
+
+def _update_state_from_reply(state: dict[str, Any], reply: str, source: str = "reply"):
```
Should be in a separate module

pantherale0 commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,395 @@
+        state["reset"] = reply
+
+
+class HegelSlowPollCoordinator(DataUpdateCoordinator):
```
Coordinator should be in its own `coordinator.py` file

pantherale0 reviewed (changes requested):
You also need to write tests for at least the config flow with 100% coverage.

boazca committed:
remove optimistic state updates - its getter home assitant out of sync when device failed to get command

boazca committed:
Move all the client logic to client file

boazca committed:
handle connection race condition

boazca committed:
handle connection reset

boazca committed:
extract client to https://github.com/boazca/hegel-ip-client

boazca committed:
Merge branch 'home-assistant:dev' into dev

boazca committed:
Merge branch 'dev' into dev

boazca committed:
add tests for config flow

joostlek commented on homeassistant/brands/hegel.json:
Brands are only created when there are more than 2 integrations

joostlek commented on homeassistant/components/hegel/__init__.py:
```diff
@@ -0,0 +1,113 @@
+        vol.Required("entity_id"): cv.entity_ids,
+        vol.Required("command"): cv.string,
+    }
+)
```
Let's  keep that for a followup

joostlek commented on homeassistant/components/hegel/__init__.py:
```diff
@@ -0,0 +1,113 @@
+
+from .const import DEFAULT_PORT, DOMAIN
+
+PLATFORMS = ["media_player"]
```
Use the `Platform` enum here

joostlek commented on homeassistant/components/hegel/__init__.py:
```diff
@@ -0,0 +1,113 @@
+        await client.start()
+        await client.ensure_connected(timeout=10.0)
+        _LOGGER.debug("Successfully connected to Hegel at %s:%s", host, port)
+    except Exception as err:
```
Don't catch bare exceptions, be more specific

joostlek commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,244 @@
+    """Config flow for Hegel amplifiers."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
```
```suggestion
```

joostlek commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,244 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class HegelFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class HegelConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,244 @@
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
```
Let's keep reconfigure for a followup PR

joostlek reviewed (changes requested):
Gave a small review
1. Please setup a proper development environment, things that should be generated are not and the code is not formatted correctly
2. The integration should have a quality_scale.yaml file and reach at least bronze

boazca committed:
handle comments from PR 153867

boazca committed:
Refactor: Split coordinator, fix types and config flow

joostlek reviewed (changes requested):
Another partial review, one thing that got me wondering

Generally the pattern with a coordinator is that you have multiple platforms, and want to centralized the polling or the data management. Currently we have the coordinator polling. Which is logical, but otoh, the media player itself is also registering to the events, why is that? Would it make more sense to keep the full state in the coordinator? What are future platforms you want to add?

boazca committed:
address comments from PR 153867

boazca commented:
About State Management & Polling Strategy:
The amplifier requires individual queries for each state (Power, Volume, Mute, Input). The integration initializes by fetching all states and subsequently relies on asynchronous update packets from the amplifier to maintain synchronization.

While continuous polling is generally unnecessary, I observed that high-frequency inputs (such as rapid volume adjustments) can occasionally result in missed packets. Therefore, the polling coordinator acts strictly as a "self-healing" failsafe rather than the primary data flow.

Current Status: This solution has been stable in local testing for the past few months. However, I am open to reducing the polling interval (e.g., to once daily) or removing it entirely—relying on the next state update to correct any drift—if a stricter event-driven approach is preferred.

pantherale0 commented:
>  The amplifier requires individual queries for each state (Power, Volume, Mute, Input). The integration initializes by fetching all states and subsequently relies on asynchronous update packets from the amplifier to maintain synchronization.

Is it worth creating a "response queue" in the library rather than immediately attempting to parse the payload? This way a separate task could handle the parsing which might unblock the listener task to prevent these missed payloads.

https://github.com/boazca/hegel-ip-client/blob/main/hegel_ip_client/client.py#L355-L366

Aside from this, personally I would remove the coordinator itself, and disable polling on the entities, but point the update function within entities to the update methods in the lib so users can use the `homeassistant.update_entity` action whenever they need to (at any interval they choose using scripts/automations).

It may also be an idea to parse the data from the amp into classes within the library, and then when you create media_player entities you can just reference the props and functions from those classes directly.

boazca commented:
Thanks for the responses.
Based on Hegel protocol, this integration support all commands so I do not expect to expand it any further.
Make sense to remove the coordinator and pass the logic on if and how often to the user.

joostlek commented:
But on the topic of commands, you originally had a service that can send commands, but we kinda don't want those as they're generally context less. In this case we do have the benefit that there's a limited set of commands (if I understood correctly). But I'd be interested what commands there are and if they can be turned into entities. As in, pressing a button entity is way more user friendly than trying to call a service and finding out what command to send and whatnot.

Also, if you feel like it, send me a message on discord :)

boazca committed:
remove coordinator.py and expose functionally through services.yaml

boazca committed:
Fix translations and tests for hegel integration

boazca committed:
Fix translations and tests for hegel integration #2

boazca committed:
Merge branch 'home-assistant:dev' into dev

boazca commented:
remove coordinator and exposed only refresh state as a service. rest of the commands are available through media player (volume change, power, input)

boazca committed:
consolidate refresh to update_entity

boazca committed:
Use client's public API instead of private attributes, add proper parent cleanup in async_will_remove_from_hass, and improve logging consistency

joostlek commented on homeassistant/components/hegel/__init__.py:
```diff
@@ -0,0 +1,81 @@
+    if unload_ok and DOMAIN in hass.data:
+        # If no more entries, clean up completely
+        if not hass.config_entries.async_entries(DOMAIN):
+            hass.data.pop(DOMAIN, None)
```
not needed

joostlek commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+
+        if unique_id:
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured()
```
why would we have none?

joostlek commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,364 @@
+            try:
+                await self._client.send(COMMANDS["input_set"](idx), expect_reply=False)
+            except (HegelConnectionError, TimeoutError, OSError) as err:
+                _LOGGER.warning("Failed to select source %s: %s", source, err)
```
If the action failed, we should raise HomeAssistantError in the case of an error from the device or ServiceValidationError when the user did something wrong

joostlek commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,364 @@
+        """
+        await super().async_will_remove_from_hass()
+
+        if self._push_handler and hasattr(self._client, "remove_push_callback"):
```
why would the client not have the remove push callback?

joostlek commented on homeassistant/components/hegel/quality_scale.yaml:
```diff
@@ -0,0 +1,36 @@
+rules:
```
Please add all rules, they don't all have to be set to done, the file needs to reflect the code

boazca commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,251 @@
+            ),
+        )
+
+    async def async_step_ssdp(
```
1. I have only h190v and I have the bug. Im not sure how wide it is but I belive its limited to h190v because its the only kind of v2 Hegel has.
2. given 1 I think pre-fill is a good middle ground

boazca commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,364 @@
+        try:
+            await self._client.send(
+                f"-r.{HEARTBEAT_TIMEOUT_MINUTES}", expect_reply=False
+            )
```
reset timeout is 3 minutes. when you send it again the timer restart.
so we do it every 2 minute, not reaching the timeout unless connection is really lost for 1-3 minutes

boazca committed:
Improve Hegel integration: use common strings, remove port config, follow HA patterns

boazca committed:
Merge branch 'dev' into dev

boazca committed:
apply prettier and pre commit validations

boazca committed:
Fix duplicate SSDP discovery entries for Hegel amplifiers

joostlek reviewed (changes requested):
I think it's going in a nice direction, there are still a few config flow comments that haven't been solved yet tho

boazca committed:
improve discovery setup

boazca committed:
fix translations

boazca committed:
fix translations #2

boazca committed:
fix test

joostlek commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,345 @@
+        entry.async_create_background_task when the config entry is unloaded.
+        """
+        await super().async_will_remove_from_hass()
+
```
Oh sorry, i meant, move the method up to async added to Hass. They're fine being separate, but when you're looking in what the entity does during its lifecycle you need to go looking for the will remove from hass method, so better to group them

boazca commented:
@joostlek I believe I address all the comments beside the logger which I didn't understand what the ideal solution

joostlek committed:
Merge branch 'dev' into boazca/dev

joostlek committed:
Remove reconfigure flow
Move async_will_remove_from_hass
Minor fixes in the flow

joostlek commented on homeassistant/components/hegel/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+                await asyncio.wait_for(
+                    asyncio.open_connection(host, DEFAULT_PORT), timeout=2.0
+                )
+            except (TimeoutError, OSError, ConnectionRefusedError) as err:
```
We should do the connection during discovery. As in, we should only propose the user to add the device when we actually are able to connect.

joostlek replied:
Sidenote, I never really worked a lot with TCP connections, but this doesn't check if this is the right device right? Why don't we use the library here?

joostlek commented on homeassistant/components/hegel/media_player.py:
```diff
@@ -0,0 +1,343 @@
+) -> None:
+    """Set up the Hegel media player from a config entry."""
+    model = entry.data[CONF_MODEL]
+    unique_id = entry.unique_id or entry.entry_id
```
For the user flow, I now added a fallback, as all SSDP discoveries have a unique id, we would fallback to entry_id for user entries

joostlek committed:
Remove reconfigure flow

joostlek reviewed (changes requested):
Okay so I had the code checked out and removed the reconfigure flow (sorry, happy to help with bringing it back after), clean some code and move some.

I think I got a good picture of what happens and I now asked a couple of questions, I think these are the final things before I will move on with the tests. As you know, feel free to DM me if you need help :)

boazca committed:
Address review comments for Hegel integration

- Add user flow for manual setup
- Use HegelClient library for connection checks instead of raw TCP
- Test connection during SSDP discovery before showing confirmation
- Use _async_abort_entries_match for user flow duplicate prevention
- Require ssdp_udn for discovery (no fallback to ssdp_usn)
- Update logger to hegel_ip_client
- Update tests for new flow structure

boazca committed:
update strings.json

joostlek committed:
Merge branch 'dev' into boazca/dev

joostlek committed:
Fix

joostlek committed:
Fix

joostlek committed:
Merge branch 'dev' into boazca/dev

joostlek committed:
Fix
